mmv1/third_party/terraform/services/storage/resource_storage_bucket.go.tmpl

```diff
@@ -549,6 +549,23 @@ func ResourceStorageBucket() *schema.Resource {
 					},
 				},
 			},
+			"hierarchical_namespace": {
```

Comment:
Will this object ever get returned by default in the future? Right now it appears not to, and only appears after getting explicitly set to true/false in its enabled field. 

If it does (even with false internal values, such as if a new field is added that causes the block to no longer serialize to null), that will break existing users and create a plan that attempts to recreate their buckets. We can mitigate, it will just cause unsetting the field to preserve the old value rather than remove the feature, i.e. turn it off.

Reply:
fyi @pavandagagoogle

Reply:
Riley this would be considered a breaking change on the API so this is not expected for the API.

Reply:
@rileykarson - Does it need to be factored in at present or is it feasible to treat  hierarchical_namespace object not being present and config with enabled=false as equivalent? This could avoid recreating the bucket which could be a valid scenario.

Reply:
As written we'd consider there to be several distinct states:

```
# 1
hierarchical_namespace = null # also unset
```

```
# 2 
hierarchical_namespace {
  enabled = false
}
```

```
# 3
hierarchical_namespace {
  enabled = true
}
```

If we can't transition from false -> null by nulling the object in an update request, we either need to apply the mitigation to consider null "preserve previous value" or add a custom one to map it onto the `enabled = false` state.

Because this field is immutable this should work as-is, but would need to be changed if:
  * update support was added in the future (to this field or any future field in the block)
  * the API changed to return a default when hierarchicalNamespace is not sent in the request at all

Reply:
In future we could have an update support for this field or any future field in the block.

So #1 & #2 should be considered as the same behavior.

Reply:
@rileykarson -  If this is something that would not be a breaking change then we can approve it else let us know what all tests we should add and what scenarios to handle.

Reply:
Discussed offline; I think the safest change is to consider null to be "preserve old value"

Reply:
Concluded to handle: null = hierachical_namespace {enabled = false } for a better UX

Reply:
Yep, let me grab a good reference for making that change!